_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,736 @@
+          authenticate_or_request_with_http_token do |token, options|          
+            ApiKey.exists?(access_token: token)
+          end
+        end
+
+      end
+    end
+  end
+```
+
+Is, well, a little awkward. Because there's really no way to 
+know who the "current_user" is without passing in some extra variable.
+
+
+#### Enter OAuth.
+
+OAuth especially via [DoorKeeper](https://github.com/doorkeeper-gem/doorkeeper) 
+and [Omniauth](https://github.com/intridea/omniauth), is amazing, 
+but popping-open new windows (or mobile apps), redirect URIs, tokens, secrets, and the funky 
+user experience that often accompany them can be a bit too much
+ceremony especially if you're connecting two *trusted* applications.
+
+#### Tokens to the rescue!
+
+Encrypting strings that can be later decrypted (usually obfuscated to a word: Token)
+is a concept oft hidden by Wizard-Of-Oz style curtains within frameworks like Rails. 
+
+Just set some random ```secret_base_key``` and you're good to go!
+
+When you dig in, oh boy, *Initialization Vectors*, *Ciphers*, wacky words / bad hacker nicknames, inspiring both dread and awe in the uninitiated. 
+But just like any good magic trick deciphering the slight of hand is both empowering and inspiring. 
+
+So let's take a peek behind the curtain.
+
+#### Imagine a scenario
+
+##### Application 1: 
+In a more traditional architecture, our Authentication service (id.wework.io for example) would be the OAuth provider.
+
+This service has a login API. For the sake of brevity, it uses [Devise](https://github.com/plataformatec/devise).
+This application also utilizes CORS via [Rack CORs](https://github.com/cyu/rack-cors) to allow appropriately configured cross-origin access.
+
+##### It has a user model
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+  # Devise password stuff
+  devise :database_authenticatable
+end
+```
+
+##### It has a base API controller
+
+```ruby
+module Api
+  class ApiController < ApplicationController  
+    skip_before_action :verify_authenticity_token
+  end
+end
+```
+
+##### It has a login controller endpoint:
+
+```ruby
+module Api
+  module V1
+    class LoginController < ApiController
+      
+      def create # POST: /api/v1/login
+        @user = User.find_by(email: params[:email])
+
+        if User.authenticate(params[:email], params[:password])
+          respond_with: @user.as_json()
+        else 
+          head 404
+        end
+      end
+
+    end
+  end
+end
+```
+
+##### It has a users controller
+
+```ruby
+module Api
+  module V1
+    class UsersController < ApiController
+      
+      def show # GET: /api/v1/users/[UUID]
+        @user = User.find_by(uuid: params[:id])
+
+        if @user.present?
+          respond_with: @user.as_json()
+        else 
+          head 404
+        end
+      end
+
+    end
+  end
+end
+```
+
+---
+
+##### Application 2: 
+Social Network (https://network.wework.io)
```

Comment:
Do we need to include a URL here? it doesn't actually go anywhere, but it is linked.